lightning/src/ln/onion_utils.rs

```diff
@@ -1258,6 +1251,165 @@ where
 	}
 }
 
+const BADONION: u16 = 0x8000;
+const PERM: u16 = 0x4000;
+const NODE: u16 = 0x2000;
+const UPDATE: u16 = 0x1000;
+
+/// The reason that a HTLC was failed by the local node. These errors either represent direct,
+/// human-readable mappings of BOLT04 error codes or provide additional information that would
+/// otherwise be erased by the BOLT04 error code.
+///
+/// For example:
+/// [`Self::FeeInsufficient`] is a direct representation of its underlying BOLT04 error code.
+/// [`Self::PrivateChanelForward`] provides additional information that is not provided by its
+///  BOLT04 error code.
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
+pub enum LocalHTLCFailureReason {
+	TemporaryNodeFailure,
+	PermanentNodeFailure,
+	RequiredNodeFeature,
+	InvalidOnionVersion,
+	InvalidOnionHMAC,
+	InvalidOnionKey,
+	TemporaryChannelFailure,
+	PermanentChannelFailure,
+	RequiredChannelFeature,
+	UnknownNextPeer,
+	AmountBelowMinimum,
+	FeeInsufficient,
+	IncorrectCLTVExpiry,
+	CLTVExpiryTooSoon,
+	IncorrectPaymentDetails,
+	FinalIncorrectCLTVExpiry,
+	FinalIncorrectHTLCAmount,
+	ChannelDisabled,
+	CLTVExpiryTooFar,
+	InvalidOnionPayload,
+	MPPTimeout,
+	InvalidOnionBlinding,
+
+	// The values below provide additional information to the BOLT04 codes enumerated above.
+	ForwardExpiryBuffer,
+	InvalidTrampolineForward,
+	PaymentClaimBuffer,
+	DustLimitHolder,
+	DustLimitCounterparty,
+	FeeSpikeBuffer,
+	ShutdownSent,
+	PrivateChannelForward,
+	RealSCIDForward,
+	ChannelNotReady,
+	InvalidKeysendPreimage,
+	InvalidTrampolineHop,
+	PaymentSecretRequired,
+	OutgoingCLTVTooSoon,
+	ChannelClosed,
+
+	/// UnknownFaliureCode represents BOLT04 failure codes that we are not familiar with. We will
+	/// encounter this if:
+	/// - A peer sends us a new failure code that LDK has not yet been upgraded to understand.
+	/// - We read a deprecated failure code from disk that LDK no longer uses.
+	UnknownFailureCode {
+		code: u16,
+	},
+}
+
+impl LocalHTLCFailureReason {
+	pub(super) fn failure_code(&self) -> u16 {
```

Comment:
these methods are _quite_ verbose, and it's annoying to have to specify the values for each enum variant twice (`failure_code` and `u16.into()`)

could try to do something [smarter](https://github.com/lightningdevkit/rust-lightning/commit/ecf39f320390c6fee0801785f527776834fa272e) here, but decided to keep it simple - interested in other opinions here!

Reply:
Could we assign the value when defining the enum using explicit discriminants?

https://doc.rust-lang.org/reference/items/enumerations.html#r-items.enum.unit-only
https://doc.rust-lang.org/reference/items/enumerations.html#r-items.enum.discriminant.explicit.unit-only